### PR TITLE
[web][3/3] Handle surrogate during line break detection

### PR DESCRIPTION
**Pull request by mdebbar**
## Description

This is one of 3 PRs to complete the line break algorithm:

1. [COMPLETE] Add the first batch of auto-generated (already passing) tests (https://github.com/flutter/engine/pull/19586).
1. [COMPLETE] Implement the algorithm and add more test cases (https://github.com/flutter/engine/pull/19610).
1. **[THIS] Fix surrogates and add the last batch of test cases**.

## Related Issues

https://github.com/flutter/flutter/issues/45932

## Tests

I added the following tests:

- `test/text/line_breaker_test.dart`

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text/line_breaker.dart [hunk 1]
```diff
@@ -91,7 +97,8 @@ bool _hasEastAsianWidthFWH(int charCode) {
 /// * https://www.unicode.org/reports/tr14/tr14-45.html#Algorithm
 /// * https://www.unicode.org/Public/11.0.0/ucd/LineBreak.txt
 LineBreakResult nextLineBreak(String text, int index) {
-  LineCharProperty? curr = _normalizeLineProperty(lineLookup.find(text, index));
+  int? codePoint = getCodePoint(text, index);
+  LineCharProperty? curr = _normalizeLineProperty(codePoint);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/line_breaker.dart:101]: Now that `_normalizeLineProperty` does not return null any more, can `curr` become non-null?

File: lib/web_ui/lib/src/engine/text/line_breaker.dart [hunk 2]
```diff
@@ -466,6 +493,24 @@ LineBreakResult nextLineBreak(String text, int index) {
...
+      if (regionalIndicatorCount.isOdd) {
+        continue;
+      } else {
+        return LineBreakResult(index, LineBreakType.opportunity);
+      }
+    }
+
+    // Do not break between an emoji base and an emoji modifier.
+    // LB30b: EB × EM
+    if (prev1 == LineCharProperty.EB && curr == LineCharProperty.EM) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/line_breaker.dart:510]: The most important line of code in this entire file 😉

File: lib/web_ui/lib/src/engine/text/line_breaker.dart (hunk 1 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:101]: Good catch! Is the analyzer supposed to warn about this?

File: lib/web_ui/lib/src/engine/text/line_breaker.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:510]: lol

**Comment by mdebbar** [comment]
I re-ran `Mac Web Engine` and it [passed](https://ci.chromium.org/p/flutter/builders/try/Mac%20Web%20Engine/6627).

Merging.

**Pull request merged**